# Hypothesis Testing

In [4]:
# Importing required packages

import statsmodels.api as sm
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import norm          # A normal continuous random variable
from numpy.random import seed         # set the randomness
from scipy import stats               # statistical operations
import warnings
warnings.simplefilter("ignore")

### One-Sample Z-Test for Mean

Let's create some dummy age data for the population of voters in the entire country and a sample of voters in North_Carolina and test whether the average age of voters in North_Carolina differs from the entire country population.

.rvs provides random samples; scale = std. deviation, loc = mean, size = no. of samples

In [7]:
# Generate Population ages data

np.random.seed(6)

population_ages1 = stats.norm.rvs(scale=18, loc=45, size=150000)
population_ages2 = stats.norm.rvs(scale=18, loc=20, size=100000)
population_ages = np.concatenate((population_ages1, population_ages2))

# Generate North Carolina sample ages data

North_Carolina_ages1 = stats.norm.rvs(scale=18, loc=40, size=30)
North_Carolina_ages2 = stats.norm.rvs(scale=18, loc=20, size=25)
North_Carolina_ages = np.concatenate((North_Carolina_ages1, North_Carolina_ages2))

print("Mean age of population of voters in the entire country: ", population_ages.mean())
print("Mean age of North Carolina population: ", North_Carolina_ages.mean())

Mean age of population of voters in the entire country:  34.97043368335889
Mean age of North Carolina population:  28.52661116077034


Notice that we used a slightly different combination of distributions to generate the sample data for North_Carolina, so we know that the two means are different. Let's conduct a z-test at a 95% confidence level and see if it correctly rejects the null hypothesis that the sample comes from the same distribution as the population.

In [11]:
# Perform Z-test for mean
one_sample_ztest = sm.stats.ztest(x1= North_Carolina_ages, x2=None, value= population_ages.mean(), alternative='smaller')

print("test-statistic (z-score): ", one_sample_ztest[0])
print("p-value: ", one_sample_ztest[1])

test-statistic (z-score):  -2.2814904209459996
p-value:  0.011259721451099942


The z-score tells us how many standard deviations from the mean our score is. In this example, our score is -2.28 standard deviations below the mean. Also, we can reject the null hypothesis that the mean of samples is equal to the population mean (i.e. 28.52) as we can see that the p-value is less than 0.05 (i.e. 0.011). This indicates that the difference between the sample mean and population mean is statistically significant.

### Two-Sample Z-Test for Mean

Let's create a sample of voters in South_Carolina and test whether the average age of voters in North_Carolina differs from the average age of voters in South_Carolina.

In [ ]:
np.random.seed(12)

# Generate South Carolina sample ages data
South_Carolina_ages1 = stats.norm.rvs(scale=15, loc=33, size=30)
South_Carolina_ages2 = stats.norm.rvs(scale=15, loc=20, size=25)
South_Carolina_ages = np.concatenate((South_Carolina_ages1, South_Carolina_ages2))

print("Mean age of South Carolina population: ", South_Carolina_ages.mean())

Mean age of South Carolina population:  24.970648445336874


In [16]:
# Perform Two sample Z-test for Mean
two_sample_ztest = sm.stats.ztest(x1=North_Carolina_ages, x2=South_Carolina_ages, value=0, alternative='two-sided')
print("test-statistic (z-score): ", two_sample_ztest[0])
print("p-value: ", two_sample_ztest[1])

test-statistic (z-score):  0.9604812179337472
p-value:  0.3368130793371993


The test yields a p-value of 0.336, which means there is a 33.6% chance we'd see the difference in the means of the two sample data are not equal to value (here, it is 0 as here we calculate the difference of the two sample means), if these two groups tested are actually identical. If we were using a 95% confidence level we would fail to reject the null hypothesis, since the p-value is greater than the corresponding significance level of 5%.

### Paired Z-Test for Mean

Consider the weights of a group of patients before and after an exercise program.

* Observation 1: The weight of a group of patients was evaluated at baseline.
* Observation 2: This same group of patients was evaluated after an 8-week exercise program.
* Variable of interest: Body weight.

In this example, we have one group with two observations, meaning that the data are paired. In this example, we know the population variance of body weights of each patient from previous studies.

The null hypothesis is that there will be no difference in weights measured before and after the exercise program. After observing that our data meet the assumptions of a paired samples z-test (the variable of interest is continuous, it has only two groups i.e. two measurements from a single group, it has paired samples, it has a normal variable of interest and the population variance is known), we proceed with the analysis.

When we run the analysis, we get a test statistic (in this case a Z-statistic) and a p-value.

The test statistic is a measure of how different the group is on our body weight variable of interest across the two observations.
The p-value is the chance of seeing our results assuming the exercise program actually doesn’t do anything. A p-value less than or equal to 0.05 means that our result is statistically significant and we can trust that the difference is not due to chance alone.

In [19]:
np.random.seed(11)

# Generate Weights of population before exercise program
before = stats.norm.rvs(scale=10, loc=50, size=100)

# Generate Weights of population after exercise program
after = before + stats.norm.rvs(scale=5, loc=-3.5, size=100)

# Create dataframe
weight_df = pd.DataFrame({"weight_before":before, "weight_after":after, "weight_change":after-before})

weight_df.head()

,weight_before,weight_after,weight_change
0,67.494547,68.365911,0.871364
1,47.139270,43.997531,-3.141739
2,45.154349,33.459091,-11.695258
3,23.466814,16.730301,-6.736513
4,49.917154,50.506002,0.588848


In [20]:
# Check summary of the data
weight_df.describe()

,weight_before,weight_after,weight_change
count,100.000000,100.000000,100.000000
mean,50.115182,46.634807,-3.480375
std,9.377513,10.423714,4.783696
min,23.466814,16.730301,-13.745286
25%,43.473681,39.362388,-6.296211
50%,50.276935,47.179677,-3.663463
75%,56.879048,53.765628,-0.511327
max,71.566744,71.336869,7.509282


The summary shows that patients lost about 3.48 pounds on average after the exercise program. Let's conduct a paired z-test to see whether this difference is significant at a 95% confidence level

In [21]:
# Perform paired Z-test
pair_ztest = sm.stats.ztest(x1=before, x2=after, value=0, alternative='two-sided')
print("test-statistic: ", pair_ztest[0])
print("p-value: ", pair_ztest[1])

test-statistic:  2.48223901119329
p-value:  0.013055966969698763


Here the null-hypothesis was that in the population (from which the samples are drawn) the difference between similarly paired observations is 0 (as indicated by value = 0). The p-value in the test output shows that the probability that the difference in means is not equal to the value (0) is 1.3%, i.e. < 5%, so the null hypothesis is rejected. The differences in means of the weight measurements before and after the exercise program is statistically significant.

### Two Sample T-test for Mean

Consider the Professional Salary Survey Results dataset. At  α  = 0.05, test whether the salary means for both male and female employees for the year 2019 who belong to the United States, are equal.

The corresponding null hypothesis and alternate hypothesis are as follows:

* $H_0$ : both salary means( male and female) are equal
* $H_A$ : both means are not equal

Now, we will carry out the two-sample t-test

In [27]:
raw_data = pd.read_csv('2019_Data_Professional_Salary_Survey_Responses.csv', header=3)
raw_data.head()

,Survey Year,Timestamp,SalaryUSD,Country,PostalCode,PrimaryDatabase,YearsWithThisDatabase,OtherDatabases,EmploymentStatus,JobTitle,...,HoursWorkedPerWeek,TelecommuteDaysPerWeek,PopulationOfLargestCityWithin20Miles,EmploymentSector,LookingForAnotherJob,CareerPlansThisYear,Gender,OtherJobDuties,KindsOfTasksPerformed,Counter
0,2017,1/5/2017 5:10:20,200000,United States,Not Asked,Microsoft SQL Server,10,MySQL/MariaDB,Full time employee,DBA,...,45,1,Not Asked,Private business,"Yes, but only passively (just curious)",Not Asked,Not Asked,Not Asked,Not Asked,1
1,2017,1/5/2017 5:26:23,61515,United Kingdom,Not Asked,Microsoft SQL Server,15,"Oracle, PostgreSQL",Full time employee,DBA,...,35,2,Not Asked,Private business,No,Not Asked,Not Asked,Not Asked,Not Asked,1
2,2017,1/5/2017 5:32:57,95000,Germany,Not Asked,Microsoft SQL Server,5,"Oracle, MySQL/MariaDB, Informix",Full time employee,Other,...,45,"None, or less than 1 day per week",Not Asked,Private business,"Yes, but only passively (just curious)",Not Asked,Not Asked,Not Asked,Not Asked,1
3,2017,1/5/2017 5:33:03,56000,United Kingdom,Not Asked,Microsoft SQL Server,6,NaN,Full time employee,DBA,...,40,1,Not Asked,Private business,"Yes, but only passively (just curious)",Not Asked,Not Asked,Not Asked,Not Asked,1
4,2017,1/5/2017 5:34:34,35000,France,Not Asked,Microsoft SQL Server,10,Oracle,Full time employee of a consulting/contracting...,DBA,...,40,"None, or less than 1 day per week",Not Asked,Private business,"Yes, but only passively (just curious)",Not Asked,Not Asked,Not Asked,Not Asked,1


In [31]:
raw_data.shape

(6893, 29)

In [32]:
len(raw_data)

6893

In [38]:
# Check for missing values
raw_data.isna().sum() / len(raw_data)

Survey Year                             0.000000
Timestamp                               0.000000
SalaryUSD                               0.000000
Country                                 0.000000
PostalCode                              0.139127
PrimaryDatabase                         0.000000
YearsWithThisDatabase                   0.000000
OtherDatabases                          0.199188
EmploymentStatus                        0.000000
JobTitle                                0.000000
ManageStaff                             0.000000
YearsWithThisTypeOfJob                  0.000000
HowManyCompanies                        0.000000
OtherPeopleOnYourTeam                   0.389961
CompanyEmployeesOverall                 0.000000
DatabaseServers                         0.000000
Education                               0.000000
EducationIsComputerRelated              0.176411
Certifications                          0.000000
HoursWorkedPerWeek                      0.000000
TelecommuteDaysPerWe

In [ ]:
# Remove missing values
raw_data.dropna(inplace=True)
raw_data.reset_index(drop= True, inplace= True)
raw_data.isna().sum()  #to check there are no more missing values

Survey Year                             0
Timestamp                               0
SalaryUSD                               0
Country                                 0
PostalCode                              0
PrimaryDatabase                         0
YearsWithThisDatabase                   0
OtherDatabases                          0
EmploymentStatus                        0
JobTitle                                0
ManageStaff                             0
YearsWithThisTypeOfJob                  0
HowManyCompanies                        0
OtherPeopleOnYourTeam                   0
CompanyEmployeesOverall                 0
DatabaseServers                         0
Education                               0
EducationIsComputerRelated              0
Certifications                          0
HoursWorkedPerWeek                      0
TelecommuteDaysPerWeek                  0
PopulationOfLargestCityWithin20Miles    0
EmploymentSector                        0
LookingForAnotherJob              

In [41]:
# Data type of SalaryUSD column
raw_data['SalaryUSD'].dtype

dtype('O')

In [42]:
# Unique values of SalaryUSD column
raw_data['SalaryUSD'].unique()

array(['200000', '95000', '47000', '41000', '51652', '63000', '45000',
       '123000', '80000', '102000', '107690', '84000', '120000', '140000',
       '138000', '40000', '72800', '99000', '50000', '55000', '175000',
       '113400', '97500', '17621', '105000', '85000', '104000', '98000',
       '70760', '67770', '82000', '98517', '107000', '78000', '72000',
       '125000', '67000', '129000', '66000', '121000', '10000', '15000',
       '130000', '87000', '70000', '35900', '61606', '96000', '27000',
       '35000', '50400', '160000', '115000', '7968', '138500', '63500',
       '88000', '111000', '134000', '95500', '100000', '122000', '26000',
       '14000', '112500', '12000', '12900', '110000', '53000', '113000',
       '90000', '93000', '94000', '65000', '92000', '120792', '145000',
       '93800', '93500', '270000', '108000', '117000', '118500', '165000',
       '73000', '109000', '62000', '70400', '153000', '116000', '149000',
       '75000', '76000', '142500', '98600', '62666', '

In [43]:
# Processing Salary column

def process_salary(salary):
    sep = '.'
    salary = str(salary)
    # Replace characters and take the cents out of our data
    salary = salary.replace(" ","").replace("$","").replace(",","").split(sep)[0]

    return float(salary)

# Replace spaces(“ “) in columns name with underscore (“_”)
raw_data.columns = raw_data.columns.str.replace(" ","_")

# Apply process_salary function
raw_data['SalaryUSD'] = raw_data['SalaryUSD'].apply(process_salary)
raw_data['SalaryUSD'].head()

0    200000.0
1     95000.0
2     47000.0
3     41000.0
4     51652.0
Name: SalaryUSD, dtype: float64

In [44]:
# Filter dataframe by year
df = raw_data[raw_data.Survey_Year == 2019]
df.head()

,Survey_Year,Timestamp,SalaryUSD,Country,PostalCode,PrimaryDatabase,YearsWithThisDatabase,OtherDatabases,EmploymentStatus,JobTitle,...,HoursWorkedPerWeek,TelecommuteDaysPerWeek,PopulationOfLargestCityWithin20Miles,EmploymentSector,LookingForAnotherJob,CareerPlansThisYear,Gender,OtherJobDuties,KindsOfTasksPerformed,Counter
1935,2019,12/6/2018 13:58:02,128500.0,United States,442,Microsoft SQL Server,15,"Microsoft SQL Server, Oracle",Full time employee,Architect,...,40,"None, or less than 1 day per week",300K-1M (large city),Private business,No,"Stay with the same employer, same role",Male,"DBA (Development Focus - tunes queries, indexe...","Manual tasks, Meetings & management, Projects",1
1936,2019,12/11/2018 6:24:30,110000.0,United States,43016,Microsoft SQL Server,18,Azure SQL DB,Full time employee,"DBA (Development Focus - tunes queries, indexe...",...,44,"None, or less than 1 day per week",1M+ (metropolis),Private business,No,"Stay with the same employer, same role",Male,DBA (Production Focus - build & troubleshoot s...,"Build scripts & automation tools, Manual tasks...",1
1937,2019,12/11/2018 6:33:56,124000.0,United States,92105,Microsoft SQL Server,18,Oracle,Full time employee,DBA (General - splits time evenly between writ...,...,40,4,1M+ (metropolis),Private business,"Yes, actively looking for something else","Stay with the same role, but change employers",Male,"DBA (Development Focus - tunes queries, indexe...","Manual tasks, Projects, R&D",1
1938,2019,12/11/2018 6:38:03,115000.0,United States,494,Microsoft SQL Server,14,"Microsoft SQL Server, Elasticsearch, Azure SQL DB",Full time employee,"DBA (Development Focus - tunes queries, indexe...",...,45,4,100K-299K (city),Private business,"Yes, actively looking for something else","Stay with the same role, but change employers",Male,DBA (Production Focus - build & troubleshoot s...,"Meetings & management, On-call as part of a ro...",1
1939,2019,12/11/2018 6:46:05,115000.0,United States,43016,Microsoft SQL Server,15,Azure SQL DB,Full time employee,"DBA (Development Focus - tunes queries, indexe...",...,55,"None, or less than 1 day per week",300K-1M (large city),Private business,"Yes, but only passively (just curious)","Stay with the same employer, same role",Male,"Architect, DBA (Production Focus - build & tro...","On-call as part of a rotation, Projects, R&D, ...",1


In [45]:
# Filter dataframe by country
US_2019 = df.loc[df.Country == 'United States', :]
US_2019.head(3)

,Survey_Year,Timestamp,SalaryUSD,Country,PostalCode,PrimaryDatabase,YearsWithThisDatabase,OtherDatabases,EmploymentStatus,JobTitle,...,HoursWorkedPerWeek,TelecommuteDaysPerWeek,PopulationOfLargestCityWithin20Miles,EmploymentSector,LookingForAnotherJob,CareerPlansThisYear,Gender,OtherJobDuties,KindsOfTasksPerformed,Counter
1935,2019,12/6/2018 13:58:02,128500.0,United States,442,Microsoft SQL Server,15,"Microsoft SQL Server, Oracle",Full time employee,Architect,...,40,"None, or less than 1 day per week",300K-1M (large city),Private business,No,"Stay with the same employer, same role",Male,"DBA (Development Focus - tunes queries, indexe...","Manual tasks, Meetings & management, Projects",1
1936,2019,12/11/2018 6:24:30,110000.0,United States,43016,Microsoft SQL Server,18,Azure SQL DB,Full time employee,"DBA (Development Focus - tunes queries, indexe...",...,44,"None, or less than 1 day per week",1M+ (metropolis),Private business,No,"Stay with the same employer, same role",Male,DBA (Production Focus - build & troubleshoot s...,"Build scripts & automation tools, Manual tasks...",1
1937,2019,12/11/2018 6:33:56,124000.0,United States,92105,Microsoft SQL Server,18,Oracle,Full time employee,DBA (General - splits time evenly between writ...,...,40,4,1M+ (metropolis),Private business,"Yes, actively looking for something else","Stay with the same role, but change employers",Male,"DBA (Development Focus - tunes queries, indexe...","Manual tasks, Projects, R&D",1


In [46]:
# Number of employees by their genders
US_2019.Gender.value_counts()

Gender
Male      170
Female     16
Name: count, dtype: int64

In [47]:
# Filter salary by gender
Male_salary = US_2019[US_2019.Gender == 'Male']['SalaryUSD']
Female_salary = US_2019[US_2019.Gender == 'Female']['SalaryUSD']

print("Mean salary for male employees: ", Male_salary.mean())
print("Mean salary for female employees: ", Female_salary.mean())

Mean salary for male employees:  107485.23529411765
Mean salary for female employees:  95481.25


In [48]:
# Perform t-test on data
ttest = stats.ttest_ind(a = Female_salary, b = Male_salary, equal_var= False)
ttest

TtestResult(statistic=np.float64(-1.2238287711542442), pvalue=np.float64(0.23553795212550016), df=np.float64(19.583191021914505))